src/Aspire.Hosting/Tools/ToolExecutionService.cs

```diff
@@ -0,0 +1,135 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Concurrent;
+using System.Runtime.CompilerServices;
+using System.Threading.Channels;
+using Aspire.Hosting.ApplicationModel;
+using Aspire.Hosting.Backchannel;
+using Aspire.Hosting.Dcp.Process;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Aspire.Hosting.Tools;
+
+internal class ToolExecutionService
```

Comment:
I think you should just run the tool inside of dcp and use the ResourceLoggingService to extract logs and ResourceNotificationService to track the lifetime of the resource.

Reply:
I think we should probably do both approaches. I think we should probably define the tool in the app model (which will support a broader range of scenarios). But `aspire exec` would effectively dynamically create an `AddExecutable(...)` call to the app model on startup.

Then its just about figuring out how you tell the orchestrator that you want resource X to effectively run with the same environment as resource Y. The way I would tackle that is that I would add an annotation which the orchestrator can spot and then make a call out to apply the target resources environment to the dynamically injected tool resource.

I kind of see these things as distinct but related things.